pkg/crd/crd.go

```diff
@@ -0,0 +1,136 @@
+)
+
+// BatchCreateCRDs will create or update the list of crds if they do not exist.
+func BatchCreateCRDs(ctx context.Context, crdClient clientv1.CustomResourceDefinitionInterface, crds []*apiextv1.CustomResourceDefinition) error {
+	existingCRDs, err := getExistingCRD(ctx, crdClient)
+	if err != nil {
+		// we do not need to return this error because we can still attempt to create the CRDs if the list fails
+		logrus.Warnf("unable to check for existing CRDs: %s", err.Error())
+	}
+
+	// ensure each CRD exist
+	for _, crd := range crds {
+		err = ensureCRD(ctx, crd, crdClient, existingCRDs)
+		if err != nil {
+			return err
+		}
+	}
+
+	existingCRDs, err = getExistingCRD(ctx, crdClient)
+	if err != nil {
+		return fmt.Errorf("unable to check for existing CRDs: %w", err)
+	}
+	// ensure each CRD is ready
+	for _, crd := range crds {
+		if existingCRD, ok := existingCRDs[crd.Name]; ok && crdIsReady(existingCRD) {
+			continue
+		}
+		// wait for the CRD to be ready
+		err := waitCRD(ctx, crd.Name, crdClient)
+		if err != nil {
+			return fmt.Errorf("failed waiting on CRD '%s': %w", crd.Name, err)
+		}
+	}
+
+	return nil
+}
+
+// getExistingCRD returns a map of all CRD resource on the cluster.
+func getExistingCRD(ctx context.Context, crdClient clientv1.CustomResourceDefinitionInterface) (map[string]*apiextv1.CustomResourceDefinition, error) {
+	storedCRDs, err := crdClient.List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to list crds: %w", err)
+	}
+
+	// convert existingCRDs to a map for faster lookup
+	existingCRDs := map[string]*apiextv1.CustomResourceDefinition{}
+	for i := range storedCRDs.Items {
+		crd := &storedCRDs.Items[i]
+		existingCRDs[crd.Name] = crd
+	}
+	return existingCRDs, nil
+}
+
+// ensureCRD checks if there is an existing CRD that matches if not it will attempt to create or update the CRD.
+func ensureCRD(ctx context.Context, crd *apiextv1.CustomResourceDefinition, crdClient clientv1.CustomResourceDefinitionInterface, existingCRDs map[string]*apiextv1.CustomResourceDefinition) error {
+	existingCRD, exist := existingCRDs[crd.Name]
+	if !exist {
+		logrus.Infof("Creating embedded CRD %s", crd.Name)
+		_, err := crdClient.Create(ctx, crd, metav1.CreateOptions{})
+		if err == nil {
+			return nil
+		}
+		if !apierrors.IsAlreadyExists(err) {
+			return fmt.Errorf("failed to create crd '%s': %w", crd.Name, err)
+		}
+		// item was not in initial list but does exist so we will attempt to get the latest resourceVersion
+		existingCRD, err = crdClient.Get(ctx, crd.Name, metav1.GetOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to get exiting '%s' CRD: %w", crd.Name, err)
+		}
+		// fallthrough and attempt and update
+	}
+	// only keep the resource version for the desired object
+	crd.ResourceVersion = existingCRD.ResourceVersion
+	// if the CRD exist attempt to update it
+	logrus.Infof("Updating embedded CRD %s", crd.Name)
+	_, err := crdClient.Update(ctx, crd, metav1.UpdateOptions{})
```

Comment:
Since you update the CRD in the cluster, would you want to also update it in your in-memory map?

Reply:
This map is not reused after this, so there is no need to update it.